File: frontend/src/types/device.ts
export type DeviceType = 'light' | 'servo' | 'fan' | 'camera' | 'sensor' | 'other';
export type DeviceTypeEnum = DeviceType;
export type DeviceModeEnum = "auto" | "manual";

export interface Device {
  device_id: number;
  device_name: string;
  device_type: DeviceType;
  pin_number: number;
  location: string;
  device_mode: DeviceModeEnum;
  status: string;
  is_active: boolean;
}

export interface DeviceCreate {
  device_name: string;
  device_type: DeviceType;
  pin_number: number;
  location: string;
  status: string;
  is_active: boolean;
}

export interface DeviceUpdate {
  device_name: string;
  device_type: DeviceType;
  pin_number: number;
  location: string;
  device_mode: DeviceType;
  status: string;
  is_active: boolean;
}

type ActionEnum =
  | "turn_on"
  | "turn_off"
  | "set_color"
  | "set_angle"
  | "set_speed";

type SourceEnum = "app" | "remote" | "auto" | "schedule";

export interface DeviceControlHistory {
  device_control_id: number;
  device_id: number;
  action: ActionEnum;
  value: string;
  source: SourceEnum;
  executed_at: string;
}

export type DeviceControlRequest = {
  device_id: number;
  action: ActionEnum;
  value: string | null;
  source: SourceEnum;
};

export interface LiveSensorData {
  temp: number;
  humi: number;
  light: number;
}

File: backend/app/schemas/device.py
from pydantic import BaseModel
from typing import Optional
from app.core.enums import DeviceTypeEnum, DeviceModeEnum

class DeviceCreate(BaseModel):
    device_name: str
    device_type: DeviceTypeEnum
    pin_number: int
    location: Optional[str] = "Unknown"
    status: Optional[str] = "online"
    is_active: Optional[bool] = True

class DeviceUpdate(BaseModel):
    device_name: Optional[str] = None
    device_type: Optional[DeviceTypeEnum] = None
    pin_number: Optional[int] = None
    location: Optional[str] = None
    device_mode: Optional[DeviceModeEnum] = None
    status: Optional[str] = None
    is_active: Optional[bool] = None

class DeviceResponse(BaseModel):
    device_id: int
    device_name: str
    device_type: DeviceTypeEnum
    pin_number: int
    location: str
    device_mode: DeviceModeEnum
    status: str
    is_active: bool
    